### ISSUE TITLE
type_repetition_in_bounds should not trigger on macro-generated code

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust-clippy#4326]
See an example from Serde in https://github.com/serde-rs/serde/issues/1595.

I consider this a false positive because in code generators we optimize for readability of the macro code, not readability of the code that gets generated.

As a simple example, a macro might want to add a where-clause:

```rust
generics.where_clause.predicates.push(parse_quote!(#param: Serialize));
```

It wouldn't make sense here to have extra logic to detect whether the where-clause in the macro input happens to already have some bounds on the same type parameter and if so then insert bounds into the same predicate, otherwise add a new predicate.

*Mentioning @xd009642, @oli-obk, @flip1995 from #3766.*
